### DISCUSSION TITLE
EF in Version Control System

### DISCUSSION THREAD

**Discussion by brettwinters** [rasmus/EventFlow#867] [Q&A]
I’m building a kind of version control system using concepts such as baseline, fork/ branch, rebase, merge, etc for a legal system 

But not sure what is the best way to use EventFlow for this.  I’ve got some elements of the following process working in EF but feel like I’m hacking the system which usually never ends well.... 

1) a “Base” AggregateRoot and Read Models which does not need public commands. All changes made thru Branch
	
2) a “Branch” AR and ReadModel (inherits from Base?) has additional properties like Base SequenceNumber (is Base’s version), BaseId, reason for change, and a List/collection of Commands or “Changes” to be implemented on Base) 
	
3) User A creates Branch from Base (via command) which gets the current version of Base and streams all Base’s events up to that point (version/sequence#) to the Branch’s ReadModel. So now the user is looking at the current state of Base 
	
4) User A executes several commands “changes” on Branch AR which emit events which are streamed to the Branch’s ReadModel but events not persisted to the event store (or alt just stored in memory since they may need to be deleted if commands change). Command added to Command/Changes List which is editable...
	
5) Note Commands in the Command /Changes List can be re-ordered, deleted, inserted, changed or the Branch can be rebased (ie new version/sequence number) and this will purge+ rebuild the read model (this is why don’t want to save events in an immutable ES) following same recess as step 3 + Commands (or their events) already in changes list (see next point about rehydrating)
	
6) Since Branch AR does not store events the state is rehydrated from Base’s Events (up to version) plus events from changes list (commands can be re-executed or we can store events and reapply them (but not in the event store) can be stored with the command). 
	
7) Finally when User A is happy (actually there is a workflow approval step) the commands or their generated events can be applied to the Base and Branch retired (thinking this can be scheduled as a job)
	
am I on the right track?

**Comment by rasmus** [comment]
If what you are trying to do is a SCM, why not simply use a SCM like git to store your data? You you could use something like [libgit2sharp](https://github.com/libgit2/libgit2sharp) I guess, I haven't done it myself.

**Comment by brettwinters** [comment]
I looked at source control initially, but my data is highly structured /complex domain model   The SCM software looks like it’s document based : 

```c#
using (var repo = new Repository("path/to/your/repo"))
{
    // Stage the file
    repo.Index.Add("file/with/changes.txt");
    repo.Index.Write();
}
```
Regarding the steps I listed - do you think it’s workable or a dead end?

**Comment by brettwinters** [comment]
adding to my previous ...I found this paper ["Maintaining Parallel Realities in CQRS and Event Sourcing"](http://engagedscholarship.csuohio.edu/cgi/viewcontent.cgi?article=1996&context=etdarchive) which talks about a command sourcing approach for branching AggregateRoots (its events), then changing (list of commands), and finally merging (execute changes or apply new events on base AR). 

> "By further applying Command Sourcing .... we present solutions to related complex milestones such as merging realities, reality optimization, conflict resolution, and aggregate duplication."

Also, Fowler talks about it ["Parallel Model - Allow an alternative representation of the state of an application, either at a different time or in a hypothetical state"](https://www.martinfowler.com/eaaDev/ParallelModel.html)
